src/overlays/actors/ovl_Bg_Haka_Ship/z_bg_haka_ship.c

```diff
@@ -41,25 +39,203 @@ const ActorInit Bg_Haka_Ship_InitVars = {
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Haka_Ship/func_8087F364.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Haka_Ship/func_8087F400.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Haka_Ship/func_8087F5C8.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Haka_Ship/func_8087F624.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Haka_Ship/func_8087F6B4.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Haka_Ship/BgHakaShip_Update.s")
+static InitChainEntry sInitChain[] = {
+    ICHAIN_VEC3F_DIV1000(scale, 100, ICHAIN_STOP),
+};
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Haka_Ship/BgHakaShip_Draw.s")
+void BgHakaShip_Init(Actor* thisx, GlobalContext* globalCtx) {
+    BgHakaShip* this = THIS;
+    s32 pad;
+    CollisionHeader* colHeader;
+
+    colHeader = NULL;
+    Actor_ProcessInitChain(&this->dyna.actor, sInitChain);
+    DynaPolyActor_Init(&this->dyna, 1);
+    this->switchFlag = (thisx->params >> 8) & 0xFF;
+    this->dyna.actor.params &= 0xFF;
+
+    if (this->dyna.actor.params == 0) {
+        CollisionHeader_GetVirtual(&D_0600E408, &colHeader);
+        this->counter = 8;
+        this->actionFunc = BgHakaShip_WaitForSong;
+    } else {
+        CollisionHeader_GetVirtual(&D_0600ED7C, &colHeader);
+        this->actionFunc = BgHakaShip_ChildUpdatePosition;
+    }
+    this->dyna.bgId = DynaPoly_SetBgActor(globalCtx, &globalCtx->colCtx.dyna, &this->dyna.actor, colHeader);
+    this->dyna.actor.world.rot.y = this->dyna.actor.shape.rot.y - 0x4000;
+    this->yOffset = 0;
+    if (this->dyna.actor.params == 0 &&
+        !Actor_SpawnAsChild(&globalCtx->actorCtx, &this->dyna.actor, globalCtx, ACTOR_BG_HAKA_SHIP,
+                            this->dyna.actor.world.pos.x + -10.0f, this->dyna.actor.world.pos.y + 82.0f,
+                            this->dyna.actor.world.pos.z, 0, 0, 0, 1)) {
+        Actor_Kill(&this->dyna.actor);
+    }
+}
+
+void BgHakaShip_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    BgHakaShip* this = THIS;
+
+    DynaPoly_DeleteBgActor(globalCtx, &globalCtx->colCtx.dyna, this->dyna.bgId);
+    func_800F89E8(&this->bellSoundPos);
+}
+
+void BgHakaShip_ChildUpdatePosition(BgHakaShip* this, GlobalContext* globalCtx) {
+    BgHakaShip* parent;
+
+    parent = (BgHakaShip*)this->dyna.actor.parent;
+    if (parent && parent->dyna.actor.update) {
+        this->dyna.actor.world.pos.x = parent->dyna.actor.world.pos.x + -10.0f;
+        this->dyna.actor.world.pos.y = parent->dyna.actor.world.pos.y + 82.0f;
+        this->dyna.actor.world.pos.z = parent->dyna.actor.world.pos.z;
+    } else {
+        this->dyna.actor.parent = NULL;
+    }
+}
+
+void BgHakaShip_WaitForSong(BgHakaShip* this, GlobalContext* globalCtx) {
+    if (Flags_GetSwitch(globalCtx, this->switchFlag)) {
+        if (this->counter) {
```

Comment:
```suggestion
        if (this->counter != 0) {
```

Reply:
I'm not really sure why but using != 0 doesn't match. My best guess is the compiler doing some optimization because of counter being unsigned